src/daft-scan/src/scan_task_iters/split_parquet_files_by_rowgroup.rs

```diff
@@ -0,0 +1,465 @@
+                        .scan_task_in_memory_size_estimate
+                        .map_or(0., |est_materialized_size| {
+                            (rg.compressed_size() as f64 / total_rg_compressed_size as f64)
+                                * est_materialized_size as f64
+                        });
+
+                    // State updates
+                    let mut new_accumulated = std::mem::take(accumulated);
+                    new_accumulated.push(*rg_idx);
+                    let new_estimated_size_bytes =
+                        *estimated_accumulated_size_bytes + rg_estimated_in_memory_size_bytes;
+                    let new_idx = idx + 1;
+
+                    // Decide which state to jump to
+                    //
+                    // If the estimated size in bytes of the materialized data is past our configured threshold, we perform a split.
+                    // NOTE: If the ScanTask is missing memory estimates for some reason, we will naively split it
+                    let scan_task_missing_memory_estimates =
+                        self.scan_task_in_memory_size_estimate.is_none();
+                    let accumulated_size_bytes_past_threshold =
+                        new_estimated_size_bytes as usize >= self.config.scan_tasks_min_size_bytes;
+                    if scan_task_missing_memory_estimates || accumulated_size_bytes_past_threshold {
+                        self.state = SplitSingleParquetFileByRowGroupsState::ReadyToEmit(
+                            new_idx,
+                            new_accumulated,
+                        );
+                    } else {
+                        self.state = SplitSingleParquetFileByRowGroupsState::Accumulating(
+                            new_idx,
+                            new_accumulated,
+                            new_estimated_size_bytes,
+                        );
+                    }
+                    continue;
+                }
+                SplitSingleParquetFileByRowGroupsState::ReadyToEmit(idx, finalized_rg_idxs) => {
+                    let num_rows = finalized_rg_idxs
+                        .iter()
+                        .map(|rg_idx| {
+                            self.file_metadata
+                                .row_groups
+                                .get(rg_idx)
+                                .unwrap()
+                                .num_rows()
+                        })
+                        .sum();
+
+                    // Create a new DataSource by mutating a clone of the old one
+                    let mut new_source = match self.scan_task.sources.as_slice() {
+                        [source] => source,
+                        _ => unreachable!(
+                            "SplitByRowGroupsAccumulator should only have one DataSource in its ScanTask"
+                        ),
+                    }.clone();
+                    if let DataSource::File {
```

Comment:
much of this can be hidden in `SplitSingleParquetFileByRowGroupsState`.

I feel like it would clearer to do something like

```
if self.state.is_ready() {
  return Some(self.state.to_scan_task())
}
```